File: badgesModifier/modifier.py
# -*- coding: utf-8 -*-

import os
import sys
from PIL import Image, ImageDraw
import xmltodict
from shutil import copyfile

colors = {
    "badge_10": 'bot.png',
    "badge_11": (241, 0, 0, 255),  # "red"
    "badge_12": (0, 193, 0, 255),  # "green"
    "badge_13": (102, 170, 255, 255),  # "blue"
    "badge_14": (204, 68, 255, 255),  # "purple"
    "badge_15": (255, 215, 0, 255),  # "golden"
    "badge_16": 'crown.png',
}

OUTPUT_PATH = './output'

def _get_base_path():
    """
    返回资源基础路径：
    - 被 PyInstaller 打包后，从临时目录 sys._MEIPASS 读取内置资源
    - 源码运行时，从当前文件所在目录读取
    """
    return getattr(sys, '_MEIPASS', os.path.abspath(os.path.dirname(__file__)))

ICON_BASE_PATH = os.path.join(_get_base_path(), 'assets')  # 图标文件基础路径

# SRC_XML = "./input/battleAtlas.xml"
# SRC_IMAGE = "./input"
# TARGET_XML = "./output/battleAtlas.xml"
# TARGET_IMAGE = "./output/battleAtlas.png"

def main():
    """
    主函数：处理雪碧图修改
    支持将指定badge替换为本地图标，其他badge保持原有圆形样式
    
    配置说明：
    - colors字典中的值可以是：
      * 颜色元组 (R, G, B, A): 绘制对应颜色的圆形
      * 图标文件名 (如 'bot.png'): 使用./input/目录下的图标文件
    
    Args:
        src_img: 源图片路径
        src_xml: 源XML配置文件路径  
        output_name: 输出文件名（不含扩展名）
    """
    src_img = sys.argv[1]
    src_xml = sys.argv[2]
    output_name = sys.argv[3]
    tg_img = os.path.join(OUTPUT_PATH, f'{output_name}.png')
    tg_xml = os.path.join(OUTPUT_PATH, f'{output_name}.xml')
    rn_img = os.path.join(OUTPUT_PATH, f'{output_name}.dds')

    atlas = xmltodict.parse(open(src_xml, "r", encoding="utf-8").read())
    coordinates = dict()
    for subTexture in atlas["root"]["SubTexture"]:
        if subTexture["name"] in colors.keys():
            coordinates[subTexture["name"]] = (
                int(subTexture["x"]),
                int(subTexture["y"]),
                int(subTexture["width"]),
                int(subTexture["height"]),
            )

    assert len(coordinates) == len(colors)

    image = Image.open(src_img)
    draw = ImageDraw.Draw(image)
    
    for badgeName, metrics in coordinates.items():
        rectangle = [metrics[0], metrics[1], metrics[0] + metrics[2], metrics[1] + metrics[3]]
        draw.rectangle(rectangle, fill=(0, 0, 0, 0))  # clear rect

        badge_config = colors[badgeName]
        
        # 检查是否为图标文件路径（字符串且以.png结尾）
        if isinstance(badge_config, str) and badge_config.endswith('.png'):
            icon_path = os.path.join(ICON_BASE_PATH, badge_config)
            
            # 如果图标文件存在，使用图标
            if os.path.exists(icon_path):
                try:
                    icon_image = Image.open(icon_path)
                    
                    # 计算图标在目标区域中的位置和大小
                    icon_width = int(metrics[2] * 0.8)  # 图标宽度为区域宽度的80%
                    icon_height = int(metrics[3] * 0.8)  # 图标高度为区域高度的80%
                    
                    # 调整图标大小
                    resized_icon = icon_image.resize((icon_width, icon_height), Image.Resampling.LANCZOS)
                    
                    # 计算居中位置
                    x_offset = metrics[0] + (metrics[2] - icon_width) // 2
                    y_offset = metrics[1] + (metrics[3] - icon_height) // 2
                    
                    # 将图标粘贴到目标位置
                    if resized_icon.mode == 'RGBA':
                        image.paste(resized_icon, (x_offset, y_offset), resized_icon)
                    else:
                        image.paste(resized_icon, (x_offset, y_offset))
                    
                    print(f"已应用图标: {badge_config} -> {badgeName}")
                except Exception as e:
                    print(f"加载图标失败 {badge_config}: {e}")
                    # 如果图标加载失败，跳过该badge
                    continue
            else:
                print(f"图标文件不存在: {icon_path}")
                # 如果图标文件不存在，跳过该badge
                continue
        else:
            # 原有的圆形绘制逻辑（颜色元组）
            dx = metrics[2] / 4
            dy = metrics[3] / 4

            rectangle[0] += dx
            rectangle[2] -= dx
            rectangle[1] += dy
            rectangle[3] -= dy
            
            if isinstance(badge_config, tuple) and len(badge_config) >= 3:
                fillColor = badge_config
                halvedAlphaColor = fillColor[:3] + (128,)
                draw.ellipse(rectangle, fill=fillColor, outline=halvedAlphaColor)
                print(f"已绘制圆形: {badgeName} -> {fillColor}")

    image.save(tg_img)
    copyfile(src_xml, tg_xml)
    os.rename(tg_img, rn_img)


if __name__ == "__main__":
    main()

File: badgesModifier/modifier.js
'use strict'

/**
 * @file badgesModifier/modifier.js
 * @description 使用 Node.js 复刻 modifier.py 逻辑：
 *  - 从 battleAtlas.xml 读取 SubTexture 坐标
 *  - 对匹配的 badge_* 区域清空并绘制：
 *      * 若配置为 PNG 图标，则缩放至 80% 并居中粘贴
 *      * 若配置为 RGBA 颜色，则绘制居中的圆形（80% 宽高）
 *  - 输出 PNG 与 XML 到 ./output，并将 PNG 重命名为 .dds
 */

const fs = require('fs')
const path = require('path')
const { promisify } = require('util')
const xml2js = require('xml2js')
const sharp = require('sharp')

const readFile = promisify(fs.readFile)
const writeFile = promisify(fs.writeFile)

const OUTPUT_PATH = path.join(__dirname, 'output')
const ICON_BASE_PATH = path.join(__dirname, 'assets')

/**
 * @type {BadgeColors}
 */
const colors = {
    badge_10: 'bot.png',
    badge_11: [241, 0, 0, 255],
    badge_12: [0, 193, 0, 255],
    badge_13: [102, 170, 255, 255],
    badge_14: [204, 68, 255, 255],
    badge_15: [255, 215, 0, 255],
    badge_16: 'crown.png',
    badge_17: 'question.png',
}

/**
 * @param {string} xmlPath
 * @returns {Promise<Array<{name:string,x:number,y:number,width:number,height:number}>>}
 */
async function parseCoordinates(xmlPath) {
    const xml = await readFile(xmlPath, 'utf8')
    // 开启 trim/normalize，并保留数组，便于统一处理
    const parsed = await xml2js.parseStringPromise(xml, { trim: true, normalize: true })
    let sub = parsed && parsed.root && parsed.root.SubTexture ? parsed.root.SubTexture : []

    // 将可能的单对象归一为数组
    if (sub && !Array.isArray(sub)) sub = [sub]

    /**
     * 递归遍历对象树，收集所有形如：
     *  - { $: { name,x,y,width,height } }
     *  - { name, x, y, width, height }（作为子元素）
     * 的节点
     * @param {any} node
     * @param {Array<any>} acc
     */
    function collectSubTextures(node, acc) {
      if (!node) return
      if (Array.isArray(node)) {
        for (const item of node) collectSubTextures(item, acc)
        return
      }
      if (typeof node === 'object') {
        const hasAttr = node.$ && (node.$.name != null) && (node.$.x != null) && (node.$.y != null) && (node.$.width != null) && (node.$.height != null)
        const hasChild = (node.name != null) && (node.x != null) && (node.y != null) && (node.width != null) && (node.height != null)
        if (hasAttr || hasChild) acc.push(node)
        for (const key of Object.keys(node)) {
          collectSubTextures(node[key], acc)
        }
      }
    }

    // 若标准位置没有取到，则回退到全树扫描
    let nodes = []
    if (Array.isArray(sub) && sub.length > 0) {
      nodes = sub
    } else {
      collectSubTextures(parsed, nodes)
    }

    function textOf(v) {
      if (v == null) return ''
      if (Array.isArray(v)) return textOf(v[0])
      if (typeof v === 'object' && v._ != null) return String(v._).trim()
      return String(v).trim()
    }

    const list = nodes
      .filter((node) => node)
      .map((node) => {
        const src = node.$ ? node.$ : node
        return {
          name: textOf(src.name),
          x: Number(textOf(src.x)),
          y: Number(textOf(src.y)),
          width: Number(textOf(src.width)),
          height: Number(textOf(src.height)),
        }
      })

    return list.filter((it) => Object.prototype.hasOwnProperty.call(colors, it.name))
}

/**
 * 在指定矩形内清空区域
 * @param {Buffer} inputPng PNG 原图数据
 * @param {{x:number,y:number,width:number,height:number}} rect
 * @returns {Promise<Buffer>} 处理后的 PNG 数据
 */
async function clearRect(inputPng, rect) {
    // 确保底图有 alpha 通道
    const base = sharp(inputPng).ensureAlpha()
    // 使用实心（alpha=1）的覆盖图，配合 dest-out 模式实现“擦除”
    const overlay = await sharp({
        create: { width: rect.width, height: rect.height, channels: 4, background: { r: 255, g: 255, b: 255, alpha: 1 } },
    }).png().toBuffer()
    return base
        .composite([{ input: overlay, left: rect.x, top: rect.y, blend: 'dest-out' }])
        .png({ compressionLevel: 3 })
        .toBuffer()
}

/**
 * 在矩形中绘制颜色圆形
 * @param {Buffer} inputPng
 * @param {{x:number,y:number,width:number,height:number}} rect
 * @param {[number,number,number,number]} rgba
 * @returns {Promise<Buffer>}
 */
async function drawCircle(inputPng, rect, rgba) {
    const circleSize = Math.floor(Math.min(rect.width, rect.height) * 0.6)

    const svg = `<svg width="${circleSize}" height="${circleSize}"><circle cx="${circleSize / 2}" cy="${circleSize / 2}" r="${circleSize / 2}" fill="rgba(${rgba[0]},${rgba[1]},${rgba[2]},${rgba[3] / 255})"/></svg>`
    const circleBuf = Buffer.from(svg)
    const left = rect.x + Math.floor((rect.width - circleSize) / 2)
    const top = rect.y + Math.floor((rect.height - circleSize) / 2)

    return sharp(inputPng)
        .composite([{ input: circleBuf, left, top }])
        .png({ compressionLevel: 3 })
        .toBuffer()
}

/**
 * 将 PNG 图标缩放为 80% 尺寸并居中贴到矩形
 * @param {Buffer} inputPng
 * @param {{x:number,y:number,width:number,height:number}} rect
 * @param {string} iconFile
 * @returns {Promise<Buffer>}
 */
async function pasteIcon(inputPng, rect, iconFile) {
    const iconPath = path.join(ICON_BASE_PATH, iconFile)
    if (!fs.existsSync(iconPath)) {
        console.warn(`图标文件不存在: ${iconPath}`)
        return inputPng
    }

    const icon = sharp(iconPath)
    const targetW = Math.floor(rect.width * 0.8)
    const targetH = Math.floor(rect.height * 0.8)
    const resized = await icon
        .resize(targetW, targetH, {
            fit: 'contain',
            background: { r: 0, g: 0, b: 0, alpha: 0 },
        })
        .png()
        .toBuffer()
    const left = rect.x + Math.floor((rect.width - targetW) / 2)
    const top = rect.y + Math.floor((rect.height - targetH) / 2)

    return sharp(inputPng)
        .composite([{ input: resized, left, top }])
        .png({ compressionLevel: 3 })
        .toBuffer()
}

/**
 * 主流程
 * @param {string} srcImg 源 PNG
 * @param {string} srcXml 贴图 XML
 * @param {string} outputName 输出名（不带扩展名）
 */
async function iconModifier(srcImg, srcXml, outputName) {
    await fs.promises.mkdir(OUTPUT_PATH, { recursive: true })
    const tgImg = path.join(OUTPUT_PATH, `${outputName}.png`)
    const tgXml = path.join(OUTPUT_PATH, `${outputName}.xml`)
    const rnImg = path.join(OUTPUT_PATH, `${outputName}.dds`)

    let imgBuf = await readFile(srcImg)
    const coords = await parseCoordinates(srcXml)

    if (coords.length !== Object.keys(colors).length) {
        throw new Error(`坐标数量(${coords.length})与 colors 键数(${Object.keys(colors).length})不一致`)
    }

    for (const c of coords) {
        const rect = { x: c.x, y: c.y, width: c.width, height: c.height }
        imgBuf = await clearRect(imgBuf, rect)
        const conf = colors[c.name]
        if (typeof conf === 'string' && conf.endsWith('.png')) {
            imgBuf = await pasteIcon(imgBuf, rect, conf)
            console.log(`已应用图标: ${conf} -> ${c.name}`)
        } else if (Array.isArray(conf) && conf.length >= 3) {
            const rgba = /** @type {[number,number,number,number]} */ (conf.length === 4 ? conf : [...conf, 255])
            imgBuf = await drawCircle(imgBuf, rect, rgba)
            console.log(`已绘制圆形: ${c.name} -> rgba(${rgba.join(',')})`)
        }
    }

    await writeFile(tgImg, imgBuf)
    await fs.promises.copyFile(srcXml, tgXml)
    await fs.promises.rename(tgImg, rnImg)
}

module.exports = { iconModifier }

